src/extensions/jana/JOmniFactory.h

```diff
@@ -0,0 +1,453 @@
+                // Otherwise this leads to a PODIO segfault
+            }
+            fac.SetCollection<PodioT>(this->collection_name, std::move(this->m_data));
+        }
+    };
+
+    void RegisterOutput(OutputBase* output) {
+        m_outputs.push_back(output);
+    }
+
+
+    // =================
+    // Handle parameters
+    // =================
+
+    struct ParameterBase {
+        std::string m_name;
+        std::string m_description;
+        virtual void Configure(JParameterManager& parman, const std::string& prefix) = 0;
+        virtual void Configure(std::map<std::string, std::string> fields) = 0;
+    };
+
+    template <typename T>
+    class ParameterRef : public ParameterBase {
+
+        T* m_data;
+
+    public:
+        ParameterRef(JOmniFactory* owner, std::string name, T& slot, std::string description="") {
+            owner->RegisterParameter(this);
+            this->m_name = name;
+            this->m_description = description;
+            m_data = &slot;
+        }
+
+        const T& operator()() { return *m_data; }
+
+    private:
+        friend class JOmniFactory;
+
+        void Configure(JParameterManager& parman, const std::string& prefix) override {
+            parman.SetDefaultParameter(prefix + ":" + this->m_name, *m_data, this->m_description);
+        }
+        void Configure(std::map<std::string, std::string> fields) override {
+            auto it = fields.find(this->m_name);
+            if (it != fields.end()) {
+                const auto& value_str = it->second;
+                if constexpr (10000 * JVersion::major
+                              + 100 * JVersion::minor
+                                + 1 * JVersion::patch < 20102) {
+                    *m_data = JParameterManager::Parse<T>(value_str);
+                } else {
+                    JParameterManager::Parse(value_str, *m_data);
+                }
+            }
+        }
+    };
+
+    template <typename T>
+    class Parameter : public ParameterBase {
+
+        T m_data;
+
+    public:
+        Parameter(JOmniFactory* owner, std::string name, T default_value, std::string description) {
+            owner->RegisterParameter(this);
+            this->m_name = name;
+            this->m_description = description;
+            m_data = default_value;
+        }
+
+        const T& operator()() { return m_data; }
+
+    private:
+        friend class JOmniFactory;
+
+        void Configure(JParameterManager& parman, const std::string& prefix) override {
+            parman.SetDefaultParameter(m_prefix + ":" + this->m_name, m_data, this->m_description);
+        }
+        void Configure(std::map<std::string, std::string> fields) override {
+            auto it = fields.find(this->m_name);
+            if (it != fields.end()) {
+                const auto& value_str = it->second;
+                if constexpr (10000 * JVersion::major
```

Comment:
I don't like the fact that we have to do this at all, and I want to delete it as soon as possible, but it's a nice trick when we are in between versions like we are right now.